cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCacheSupplier.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.messaging;
+
+import co.cask.cdap.data2.transaction.coprocessor.CacheSupplier;
+import co.cask.cdap.data2.transaction.queue.hbase.coprocessor.CConfigurationReader;
+import co.cask.cdap.data2.util.ReferenceCountedSupplier;
+import co.cask.cdap.data2.util.hbase.ScanBuilder;
+import com.google.common.base.Supplier;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessorEnvironment;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Supplies instances of {@link TopicMetadataCache} implementations.
+ */
+public class TopicMetadataCacheSupplier implements CacheSupplier<TopicMetadataCache> {
+  private static AtomicReference<TopicMetadataCache> instance = new AtomicReference<>();
+  private static AtomicInteger refCount = new AtomicInteger(0);
+  private static final Object lock = new Object();
+
+  private final ReferenceCountedSupplier<TopicMetadataCache> referenceCountedSupplier;
```

Comment:
If `referenceCountedSupplier` instance is made static then we can move `instance`, `lock` and refCount` as instance variable of class `ReferenceCountedSupplier`, right?